### ISSUE TITLE
🚀 Feature: Database transactions

### ISSUE THREAD

**Issue by locohost** [appwrite/appwrite#2788]
### 🔖 Feature description

Really need to be able to group related create/read/update/delete operations into a database transaction so we can rollback everything on any failure.

### 🎤 Pitch

Without database transactions, data integrity will very difficult to maintain.

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/appwrite/appwrite/blob/HEAD/CODE_OF_CONDUCT.md)

**Comment by eldadfux** [comment]
We do have this in mind, any more community feedback about this feature would be helpful. 

Some of my thoughts are to allow one of two new API endpoints:

`POST v1/database/collections/:id/transaction` - this will accept an array of documents from the same collection and execute them as a single transaction.

or

`POST v1/database/transaction` - this could receive a set of documents from multiple collections and execute them as a transaction, much more flexible, but might be a bit tricky to provide information about what document is owned by what collection which can make the payload a bit complex to build.

**Comment by JakeAi** [comment]
> We do have this in mind, any more community feedback about this feature would be helpful.
> 
> Some of my thoughts are to allow one of two new API endpoints:
> 
> `POST v1/database/collections/:id/transaction` - this will accept an array of documents from the same collection and execute them as a single transaction.
> 
> or
> 
> `POST v1/database/transaction` - this could receive a set of documents from multiple collections and execute them as a transaction, much more flexible, but might be a bit tricky to provide information about what document is owned by what collection which can make the payload a bit complex to build.

I had the thought of creating a cookie/token and using that cookie/token as the transaction id. Any request that uses that token is part of the transaction. You would need to call an endTransaction function to complete the session. Transactions timeout after a few seconds of no activity?

**Comment by locohost** [comment]
I would say that allowing transactions on a single collection would be "ok" and might have some uses, however it's not a real-world scenario. Our real transactions will involve creating and updating documents in multiple different collections of linked documents. These all need to succeed or else we have to rollback.

**Comment by locohost** [comment]
I like @JakeAi idea above a lot. That's a pretty standard transaction.

**Comment by JakeAi** [comment]
> I like @JakeAi idea above a lot. That's a pretty standard transaction.

Yeah, I like it too LOL. Thanks

**Comment by eldadfux** [comment]
I like it 3.

**Comment by JakeAi** [comment]
If there is a large amount of data on a slow network. you could add a timeout to startTransaction. 5 seconds after the last request would cancel the transaction. 
```TS
    const transaction = this.appwriteDocumentsService.startTransaction(5000); // returns new database object with transaction header set
    transaction.createDocument();
    transaction.createDocument();
    transaction.createDocument();
    transaction.deleteDocument();
    transaction.updateDocument();
    transaction.updateCollection();
    transaction.createCollection();
    transaction.endTransaction();
```

**Comment by locohost** [comment]
> Hi, i think the transaction should not be limited to document collections.
> 
> I typically create documents, files and teams in same time.

Yeah that's a good point. Agree. But it would be really easy to remove a file or team when the transaction fails. It's a work-around. Assuming we at least have collection transactions I mean :-)

**Comment by anandsubbu007** [comment]
is this feature available now?

**Comment by damien-o** [comment]
May be another a way to address this problem would be to write something like that :
```
var del = {
    collection: 'ColX',
    type: Action.Delete,
    id: '123',
    version: 456, // timestamp
};

var insert = {
    collection: 'ColX',
    type: Action.Insert,
    data: { prop: 'value', ....}
};

var update = {
    collection: 'ColX',
    type: Action.Update,
    id: '123',
    version: 456, // timestamp,
    data: { prop: 'value', ....}
};


var count = await this.appwriteDocumentsService.execute([del, insert, update]);
```

This avoids the need to maintain the transaction open and wait until commit or rollback is called.
In addition the row version, ensures that no other processes modified the data since you read it,

A more flexible way could be :
```
var update = {
    collection: 'ColX',
    type: Action.Update,
    predicate: {
      id: '123',
      version: 456
    },
    data: { prop: 'value', ....}
};
```

**Comment by JakeAi** [comment]
> May be another a way to address this problem would be to write something like that :
> 
> ```
> var del = {
>     collection: 'ColX',
>     type: Action.Delete,
>     id: '123',
>     version: 456, // timestamp
> };
> 
> var insert = {
>     collection: 'ColX',
>     type: Action.Insert,
>     data: { prop: 'value', ....}
> };
> 
> var update = {
>     collection: 'ColX',
>     type: Action.Update,
>     id: '123',
>     version: 456, // timestamp,
>     data: { prop: 'value', ....}
> };
> 
> 
> var count = await this.appwriteDocumentsService.execute([del, insert, update]);
> ```
> 
> This avoids the need to maintain the transaction open and wait until commit or rollback is called. In addition the row version ensures that no other processes modified the data since you read it,
> 
> A more flexible way could be :
> 
> ```
> var update = {
>     collection: 'ColX',
>     type: Action.Update,
>     predicate: {
>       id: '123',
>       version: 456
>     },
>     data: { prop: 'value', ....}
> };
> ```

Seems very unnatural and unnecessary.

**Comment by meepeek** [comment]
Is it a good idea to make transaction a specific option on creating collection ?

I think making transaction available in general comes many sacrifices and most of the time, we don't need it. That's why noSQL concept flourish recently by trade off the feature for performance and scalability.

Conclusion of the idea is:
1. Collections will have 2 categories, transaction and non-transaaction
2. A sequence sensitive collection will require transaction flag on creation
3. every insert and update operation of the transaction collection will go thru a collection dedicated, queue management micro service. Non-transaction collection will skip this so it would be more performant.

Please note that I'm not expert in the field, this is just an opinion for the knowledge I have so far. Also, I don't have experience for programming in cluster. Feels free to point out for any flaw in the idea.

**Comment by merabtenei** [comment]
@meepeek How do you solve the race condition problem without transactions ? Let's say we have a post likes counter you want to increment every time a user likes the post. Without transactions this is vulnerable to race condition, and with enough traffic and two events at the same time we can easily get wrong values.
I didn't find any solution for this with AppWrite right now, Firebase has transactions and also the FieldValue.increment as a parameter to update.
I was very excited to start a side project with AppWrite but this hit me on the face. I guess AppWrite needs a bit more features to be mature enough for production.

**Comment by ddenev** [comment]
Dear @eldadfux 

as a leader of this project, would you be kind to shine light on your plans for implementing this request? Do you have a roadmap for adding transactions, it this a priority, are you working on this, etc.

thank you!

**Comment by eldadfux** [comment]
Hey @ddenev. Yes this is a priority and we do plan to work on this feature. We want to include in one of the next two database iterations, but still hard to tell when exactly could the community expect it.

**Comment by tannermeade** [comment]
@eldadfux thank you for the update on transactions. This is definitely a high priority feature for any serious backend involving payments or most enterprise level use cases.

Having a single transaction across multiple collections is the use case that we would need. The use case involves payments, ownership of products, and registration of those products which is not possible to represent in a single collection.

**Comment by Panth977** [comment]
I have a simple solution,


> > We do have this in mind, any more community feedback about this feature would be helpful.
> > Some of my thoughts are to allow one of two new API endpoints:
> > `POST v1/database/collections/:id/transaction` - this will accept an array of documents from the same collection and execute them as a single transaction.
> > or
> > `POST v1/database/transaction` - this could receive a set of documents from multiple collections and execute them as a transaction, much more flexible, but might be a bit tricky to provide information about what document is owned by what collection which can make the payload a bit complex to build.
> 
> I had the thought of creating a cookie/token and using that cookie/token as the transaction id. Any request that uses that token is part of the transaction. You would need to call an endTransaction function to complete the session. Transactions timeout and do not execute after a few seconds of no activity?
> 
> ```ts
>     const transaction = this.appwriteDocumentsService.startTransaction(); // returns new database object with transaction header set
>     transaction.createDocument();
>     transaction.createDocument();
>     transaction.createDocument();
>     transaction.deleteDocument();
>     transaction.updateDocument();
>     transaction.updateCollection();
>     transaction.createCollection();
>     transaction.endTransaction();
> ```

as @JakeAi wrote, i completely agree
to make this easier for appwrite developer & appwrite users, we can do the following

```ts
const changesOn = [{database, collection, document?}] // reference to all the collection Or documents developer wants to lock while transaction is active
// incoming requests on this ref (write changes), collections/documents will either be denied Or remain pending if they are in lock state.
 
const transaction = this.appwriteDocumentsService.startTransaction(5000, changesOn); // will exit in 5000ms if not already & will put all changesOn reference under lockdown
    transaction.createDocument(); // creation can be allowed for any collection
    transaction.createDocument(); 
    transaction.createDocument();
    transaction.deleteDocument(); // deletion & update is allowed only if changesOn contains the reference to the document
    transaction.updateDocument(); 
    transaction.updateCollection(); // update on collection is allowed if changesOn contains the reference to the collection (without document specification)
    transaction.createCollection(); // creation can be allowed for any collection
    transaction.endTransaction(); // will end the lockdown for changesOn reference
```

**Comment by fogelito** [comment]
Hello everybody, 
Wanted to comment regarding transactions , that they happen only for on insert/update/delete executions.
So to have transaction.createCollection(); for example will probably be impossible since altering the data structure will make an auto-commit execution that will commit what's in the transaction and end it.

**Comment by Yavnik** [comment]
Dear @eldadfux  @fogelito,

Thank you for addressing this issue.
I really love Appwrite and all it's features but not having Transactions in DB is kind of a deal breaker for my project.

Do you have any update on it's implementation or a timeline?

**Comment by eldadfux** [comment]
I can finally confirm that work on relationships / nested documents has officially began! Transaction are probably one of the next features just after that.

**Comment by nev-21** [comment]
Thank you @eldadfux, please this is an important feature

**Comment by nesterow** [comment]
I am also going to need transaction support. 
I am thinking, is there a way to use some ORM in a function? Or maybe a way to create endpoints that execute raw SQL?

**Comment by danilo73r** [comment]
Based on https://github.com/appwrite/appwrite/discussions/4957#discussioncomment-4605291
It is not a god idea to directly update db data cause cache might get out of sync.
Only directly reads to the db would be ok.

**Comment by danilo73r** [comment]
> `POST v1/database/transaction` - this could receive a set of documents from multiple collections and execute them as a transaction, much more flexible
I think this is a good approach to db transactions. 

@eldadfux Please consider to make it more flexible, it would be great if commit and rollback were accessible to developers.
For example, in "v1/database/transaction" a param "behaviour" can be 'default' or 'manual'. 
- Default will manage the commit/rollback of the transaction internally (The developer only have to call v1/database/transaction). 
- And manual will requiere to call other sdk function/endpoint to commit or rollback with the transaction id (which was obtained from v1/database/transaction).

**Comment by aabmets** [comment]
As a workaround until this feature is integrated into Appwrite, it is possible to create a collection of session documents and implement a crude ORM on top of Appwrite SDK which collects all DB operations into a single data block and sends that to a backend function which parses all of the db operations and creates equivalent reversal operations for each of the db queries and then attempts to execute each of the db queries one-by-one and track their operation success. If an exception is raised, all of the reversal queries up to the failed query are executed. This approach does not protect from internal Appwrite failure, but it will protect against network connectivity issues between the client app and the backend.

**Comment by lisu-wefind-ai** [comment]
Hi dear Members,
May I get this feature's new progress?
Will it be in the roadmap?
Thanks very much!

**Comment by nev-21** [comment]
Hi, it would be great if the transaction isolation level can be specified too as a parameter when a trasaction start. 
And update method (for documents api) have the "queries" parameter, that will solve too much problems

**Comment by tbsvttr** [comment]
A transaction feature would be very important to streamline development for projects that are not trivial.

**Comment by LDY681** [comment]
This is one of the key features to make Appwrite production ready. Hope to see it implemented soon.

**Comment by AmitGurbani** [comment]
@eldadfux When can we expect this feature?

**Comment by PratikDev** [comment]
> Hi, i think the transaction should not be limited to document collections.
> 
> I typically create documents, files and teams in same time.
> 
> ```
>     transaction.createDocument();
>     transaction.createDocument();
>     transaction.createDocument();
>     transaction.deleteDocument();
>     transaction.updateDocument();
>     transaction.updateCollection();
>     transaction.createCollection();
>     transaction.createFile();
>     transaction.updateFile();
>     transaction.deleteFile();
>     transaction.createTeam();
>     transaction.updateTeam();
>     transaction.deleteTeam();
>     ...
>     transaction.endTransaction();```
> ```

I agree with @aleslanik2. there can be cases where a transaction needs to create/update/delete a doc and also upload/update/delete a file.
i.e. post creation in a social media app. The user can upload multiple files also with some texts. so if any of the files fail to upload, we don't want the post to be created in our doc. also, we don't want to keep the files that were uploaded successfully in our storage

**Comment by aabmets** [comment]
This feature issue feels like it's going to end up like a mozilla bug that is finally resolved 13 years in the future by an undergraduate intern. :P

**Comment by anatoliykmetyuk** [comment]
Currently the absence of this feature is the main decisive factor for me why I can't switch from firebase.

**Comment by it-delta** [comment]
I can't imagine a production work without transactions.
It's one of the key point of databases.

**Comment by FernandoUFS** [comment]
Yeah, I have functions that create collections and their attributes on AppWrite cloud. Sometimes some attribute fail to be created. And I end up with broken collections. With transactions I'd be able to rollback, and try again.

**Comment by SanjivJais** [comment]
I've already developed a production level project and almost ready to deploy it. But my major concern is the transactions feature. My project deals with in-app currencies and in case of any failure, users could get free money or lose their money for nothing. This is a big issue.

**Comment by RoyMatipa** [comment]
> I've already developed a production level project and almost ready to deploy it. But my major concern is the transactions feature. My project deals with in-app currencies and in case of any failure, users could get free money or lose their money for nothing. This is a big issue.   

I'm in the same situation

**Comment by aabmets** [comment]
> I've already developed a production level project and almost ready to deploy it. But my major concern is the transactions feature. My project deals with in-app currencies and in case of any failure, users could get free money or lose their money for nothing. This is a big issue.

@eldadfux

**Comment by Khalinkara** [comment]
I'm waiting for this feature so I can migrate from Pocketbase to Appwrite. I can't rely on financial data without transactions

**Comment by cccabdalla** [comment]
@eldadfux 
Is transactions feature done so far ?
I really need and i thought appwrite had this in place already but ended up reading the requests submitted to you for this important feature but no updates so far !

**Comment by aabmets** [comment]
Appwrite uses internally [MariaDB](https://mariadb.com/kb/en/transactions/), which by default supports transactions, so it is puzzling to me why this feature was not implemented into Appwrite from version 1.0

**Comment by MeisterDerMagie** [comment]
What is the status here? This feature is extremely important for a professionally used database.

**Comment by Mehdi-Verfaillie** [comment]
This feature request is the most upvoted one, originally submitted 3 years ago. Without the green light from the core maintainers, it will never be implemented. Appwrite is great, but without this and proper DB schema migration tool, it will remain a tool suited for side projects and small-scale applications rather than enterprise-level projects.

**Comment by merabtenei** [comment]
> This feature request is the most upvoted one, originally submitted 3 years ago. Without the green light from the core maintainers, it will never be implemented. Appwrite is great, but without this and proper DB schema migration tool, it will remain a tool suited for side projects and small-scale applications rather than enterprise-level projects.

Totally agree, I love appwrite but I only used it for really small projects because it doesn't fit with the requirements of enterprise project. We looked at it when we started our startup but we ended going with the old but reliable tech Python Django and PostgreSQL

**Comment by LeonardoRick** [comment]
Sad to have been tracking this for two years and still not be able to perform transactions. I’m sure it’s a complex feature to add, but I can’t think of anything that would bring more value to the library than implementing this.

**Comment by Tomic-Riedel** [comment]
Hey @stnguyen90 & @eldadfux,
could we please get an official update on when we can expect database transactions?

It’s been in the 1.7 Release project for so long and is one of the most upvoted issues in the repo, so I think a lot of people were expecting some kind of announcement at Init. It’s honestly pretty frustrating to be blocked from using a tool you usually love—not because of a nice-to-have feature, but something that’s essential—and then not even get any communication around when it might be released.

I totally get that this is a big feature to implement. But removing it from the 1.7 milestone without any update leaves a lot of people hanging, and that’s hard to understand.

Would really appreciate some clarity here!

**Comment by sanny-io** [comment]
I'm in the same position as a lot of other people here. 

I can't use Appwrite's database for anything production related until this is done 😢

**Comment by user425846** [comment]
Am i seeing it right, that the last official reply in this issue was 2 years ago and it was just removed a week ago from the 1.7 Release? Without ANY information or statement why or how much longer it will take? This is absolutely unacceptable and just shows that appwrite is not production ready yet. I will not be using appwrite for my next project.

**Comment by tessamero** [comment]
Hey all! I really appreciate all of your continued feedback and engagement on this issue. It's very clear that database transactions are a very important feature for many, especially for production and enterprise-level use cases.

An update from the engineering team: An internal RFC for database transactions has been complete and approved and created a draft PR.  We will still need to do some testing, benchmarking, and addressing any edge cases. We have a rough estimate of a few weeks, depending on how things go with QA and feedback.

We're also potentially planning on making the RFC public soon so we can gather feedback from the community before finalizing everything.

Thank you so much for being patient!

**Comment by mariusbolik** [comment]
CSV Imports are now available in Appwrite: https://www.youtube.com/watch?v=7UfWaHNghas

**Comment by apperside** [comment]
When I found appwrite I felt like I discovered the holy grail, but I was so sad to find that there is no support for transactions, and working with a database without them is like driving a lamborghini without breaks. MUST have

**Comment by JakeAi** [comment]
It’s about the money unfortunately

**Comment by hadnet** [comment]
> Hey all! I really appreciate all of your continued feedback and engagement on this issue. It's very clear that database transactions are a very important feature for many, especially for production and enterprise-level use cases.
> 
> An update from the engineering team: An internal RFC for database transactions has been complete and approved and created a draft PR.  We will still need to do some testing, benchmarking, and addressing any edge cases. We have a rough estimate of a few weeks, depending on how things go with QA and feedback.
> 
> We're also potentially planning on making the RFC public soon so we can gather feedback from the community before finalizing everything.
> 
> Thank you so much for being patient!

Any updates?

**Comment by Aristides-19** [comment]
Are there any updates on this?

**Comment by user425846** [comment]
Lets wait another 3 years until they give us an update again. Im setting up Supabase right now. Its crazy how such an essential feature like database transactions is still not available. Appwrite is just not ready for anything beyond super simple apps.

**Comment by PratikDev** [comment]
I subscribed to this issue exactly 2 years back, hoping that it'd be solved soon. I'm unsubscribing today

**Comment by cccabdalla** [comment]
Also wondering why Appwrite has no desire to reply to us !!!

**Comment by user425846** [comment]
> Also wondering why Appwrite has no desire to reply to us !!!

Especially because this is, by far, the most upvoted issue.

**Comment by tessamero** [comment]
Hey all, Tessa from the Appwrite team. Database Transactions is one of the most requested features and critical for production use. The good news is that it's NOT on hold. Our engineering team is actively working on it now and it's one of our top priorities. We are making sure it's solid before we can release it. Thanks for your patience. To get updates on the features we have been releasing weekly, you can join our discord (https://appwrite.io/discord) and see the announcements channel, or follow X (https://x.com/appwrite)  :)

**Comment by cccabdalla** [comment]
Great to hear from you, Appwrite team. Thank you.

**Comment by user425846** [comment]
> Hey all, Tessa from the Appwrite team. Database Transactions is one of the most requested features and critical for production use. The good news is that it's NOT on hold. Our engineering team is actively working on it now and it's one of our top priorities. We are making sure it's solid before we can release it. Thanks for your patience. To get updates on the features we have been releasing weekly, you can join our discord (https://appwrite.io/discord) and see the announcements channel, or follow X (https://x.com/appwrite) :)

Thanks for the reply. Is there ANY kind of ETA on this? We really need to know.

**Comment by hadnet** [comment]
@tessamero and what about #7369 ? You guys can’t just drop the "beta" tag and leave this core feature behind! Query relationships are a must-have! Come on, seriously!

**Comment by tessamero** [comment]
> > Hey all, Tessa from the Appwrite team. Database Transactions is one of the most requested features and critical for production use. The good news is that it's NOT on hold. Our engineering team is actively working on it now and it's one of our top priorities. We are making sure it's solid before we can release it. Thanks for your patience. To get updates on the features we have been releasing weekly, you can join our discord (https://appwrite.io/discord) and see the announcements channel, or follow X (https://x.com/appwrite) :)
> 
> Thanks for the reply. Is there ANY kind of ETA on this? We really need to know.

It'll be pretty soon, the engineering team is working out some issues and ensuring it works smoothly before releasing it. Hard to make an exact date eta but we are estimated a couple weeks ish. It was supposed to be sooner but had other major priority fix/updates for our small team of engineers. We are growing our engineering team and hiring more people to help ship features a bit faster. :) Trust me, it's an important feature even our team members want to use, so we are doing our best to ensure it happens.

**Comment by camorobot** [comment]
Love the work you're guys are doing!! any updates on the estimated release data?

**Comment by tessamero** [comment]
It is going to be one of our next announcements this month :) the team is
working out the final issues and should be good to go!

**Comment by aabmets** [comment]
@tessamero It would be funny self-deprecating humour if Appwrite awarded itself the "Longest Open Feature Request Issue" award each year.

**Comment by cccabdalla** [comment]
Any updates so far?

**Comment by user425846** [comment]
It worries me a bit that there is a appwrite 1.8 RC release without any mention of transactions

**Comment by camorobot** [comment]
@user425846  Yes, I agree. I haven't come across anything that mentions transactions either. Perhaps this could be clarified? Or maybe I overlooked it, in which case I apologize.

**Comment by tessamero** [comment]
Thanks for hanging in there on this thread. Transactions are now available!

Learn more: https://appwrite.io/blog/post/announcing-transactions-api

We appreciate all the input. please try it out and open a fresh issue if you hit any edge cases.

**Comment by SanjivJais** [comment]
Finally Appwrite has released the Transactions API 🎉 Read the [full announcement here](https://appwrite.io/blog/post/announcing-transactions-api)

**Comment by cccabdalla** [comment]
Just wow !

Thank you so much.
